docs/roslyn_best_practices.md

```diff
@@ -0,0 +1,181 @@
+
+**Should**
+* Prefer `Task<Document>` when possible and use `Task<Solution>` as needed when returning a result from a Code Fixer.
+
+### 6. Use SolutionEditor and DocEditor when you need to batch changes
+When you need to make numerous changes you will want to batch those changes with the SolutionEditor, or DocEditor. It can be helpful to think about these objects as you would think about using StringBuilder when concatenating a large number of strings. These objects enable you to create a cumulative list of changes that can reduce GC pressure created by modifying immutable SyntaxTrees.
+
+If you’re not batching changes, then most code changes can be accomplished by working directly with the document’s syntax root and the SyntaxGenerator.
+
+
+### 7. Enable support for Fix All
+Don’t force users to fix instances of a diagnostic one-by-one. A FixAll occurrences code fix means: I have a code fix 'C', that fixes a specific instance of diagnostic 'D' in my source and I want to apply this fix to all instances of 'D' across a broader scope, such as a document or a project or the entire solution.
+
+Your code fixer should override `GetFixAllProvider` to return a non-null instance of [FixAll Provider](https://github.com/dotnet/roslyn/blob/main/docs/analyzers/FixAllProvider.md).
+
+### 8. Treat the Code Fixer as a single unit of work
+Code fixers are applied as a one-step change and their order is not guaranteed. If the code you’re adding requires a namespace import then your code fixer should perform that change if necessary. Do not assume that another code fixer will be responsible for adding the correct namespace.
+
+## Considerations specific to upgrade-assistant
+
+There are many stopping points on the journey from .NET Framework to .NET latest and even more scenarios specific to our customers. Because of this, `upgrade-assistant` does not always assume that the code can be compiled. This section highlights some considerations specific to the "work in progress" state that all projects will pass through as they are upgraded.
+
+### 1. Do not assume the code will compile
+`upgrade-assistant` will make sweeping changes across the solution. These sweeping changes often result in manual changes that must be made after running upgrade assistant before the code will compile.
+
+As an example, `upgrade-assistant` will upgrade NuGet packages across major versions. `upgrade-assistant` does not evaluate if the newer package contains breaking changes.
+
+### 2. Symbols may not be resolvable
+`upgrade-assistant` will add, upgrade, and remove package references during the upgrade process. If your analyzer is looking for a specific symbol then you may need to [AddMetadataReference](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.project.addmetadatareference) to ensure the symbol is available.
+
+### 3. Add reference assemblies instead of the assembly implementation
```

Comment:
I bet we could come up with a way to have analzyer declare what reference assemblies they require to simplify this

Reply:
Potentially, I'm not sure we actually want reference assemblies. Adding them certainly increases accuracy but I think we are able to reach a reasonable level of certainty without the reference.

Having said that, I wonder if we care about AddMetadataReference